File: sdks/node/src/plugins/qwen-config.ts
import type { JSONPath } from 'jsonc-parser';
import type { FileSnapshot } from './managed-files.js';
import { changeJsonc } from './jsonc-edit.js';
import {
  MANAGED_EVENTS,
  managedQwenRemovals,
  readQwenHooks,
  type ManagedQwenEvent,
  type QwenGroup,
  type QwenHooks,
} from './qwen-contract.js';
import { isObject, parseCommentedJsonObject, type JsonObject } from './strict-json.js';

export const QWEN_OWNED_FILE_MARKER = '// Managed by Elydora';

export type QwenDocumentKind = 'system-defaults' | 'user' | 'workspace' | 'system';

export interface QwenDocument {
  readonly kind: QwenDocumentKind;
  readonly filePath: string;
  readonly exists: boolean;
  readonly raw: string;
  readonly snapshot?: FileSnapshot;
  readonly root: JsonObject;
  readonly hooks: QwenHooks;
  readonly hasHooksContainer: boolean;
  readonly disableAllHooks?: boolean;
  readonly folderTrustEnabled?: boolean;
  readonly ownedFile: boolean;
}

export interface RenderedQwenDocument {
  readonly document: QwenDocument;
  readonly changed: boolean;
  readonly next?: string;
}

interface DocumentOptions {
  readonly kind: QwenDocumentKind;
  readonly exists: boolean;
  readonly filePath: string;
  readonly raw: string;
  readonly snapshot?: FileSnapshot;
}

export function qwenSourceLabel(kind: QwenDocumentKind): string {
  switch (kind) {
    case 'system-defaults': return 'Qwen Code system defaults';
    case 'workspace': return 'Qwen Code workspace settings';
    case 'system': return 'Qwen Code system override settings';
    default: return 'Qwen Code user settings';
  }
}

function readFolderTrustEnabled(root: JsonObject, label: string): boolean | undefined {
  if (root.security === undefined) return undefined;
  if (!isObject(root.security)) throw new Error(`${label} field "security" must be an object`);
  if (root.security.folderTrust === undefined) return undefined;
  if (!isObject(root.security.folderTrust)) {
    throw new Error(`${label} field "security.folderTrust" must be an object`);
  }
  const enabled = root.security.folderTrust.enabled;
  if (enabled !== undefined && typeof enabled !== 'boolean') {
    throw new Error(`${label} field "security.folderTrust.enabled" must be a boolean`);
  }
  return enabled as boolean | undefined;
}

export function parseQwenDocument(options: DocumentOptions): QwenDocument {
  const label = `${qwenSourceLabel(options.kind)} at ${options.filePath}`;
  const root = parseCommentedJsonObject(options.raw, label);
  if (root.disableAllHooks !== undefined && typeof root.disableAllHooks !== 'boolean') {
    throw new Error(`${label} field "disableAllHooks" must be a boolean`);
  }
  const hasHooksContainer = Object.hasOwn(root, 'hooks');
  return {
    ...options,
    root,
    hooks: readQwenHooks(root.hooks),
    hasHooksContainer,
    disableAllHooks: root.disableAllHooks as boolean | undefined,
    folderTrustEnabled: readFolderTrustEnabled(root, label),
    ownedFile: options.kind === 'user' && options.raw.startsWith(QWEN_OWNED_FILE_MARKER),
  };
}

export function createQwenDocument(
  kind: QwenDocumentKind,
  filePath: string,
): QwenDocument {
  return parseQwenDocument({
    kind,
    exists: false,
    filePath,
    raw: kind === 'user' ? `${QWEN_OWNED_FILE_MARKER}\n{}\n` : '{}\n',
  });
}

function currentDocument(document: QwenDocument, raw: string): QwenDocument {
  return parseQwenDocument({
    kind: document.kind,
    exists: document.exists,
    filePath: document.filePath,
    raw,
    snapshot: document.snapshot,
  });
}

function removeManagedEntries(
  document: QwenDocument,
  raw: string,
  agentId?: string,
): string {
  const removals = managedQwenRemovals(document.hooks, agentId);
  for (const event of MANAGED_EVENTS) {
    const eventRemovals = removals
      .filter((removal) => removal.event === event)
      .sort((left, right) => right.groupIndex - left.groupIndex);
    for (const removal of eventRemovals) {
      const groupPath: JSONPath = ['hooks', event, removal.groupIndex];
      if (removal.removeGroup) {
        raw = changeJsonc(raw, groupPath, undefined);
        continue;
      }
      for (const handlerIndex of [...removal.handlerIndexes].sort(
        (left, right) => right - left,
      )) {
        raw = changeJsonc(raw, [...groupPath, 'hooks', handlerIndex], undefined);
      }
    }
    if (eventRemovals.length > 0) {
      const current = currentDocument(document, raw);
      if ((current.hooks[event] ?? []).length === 0) {
        raw = changeJsonc(raw, ['hooks', event], undefined);
      }
    }
  }
  const current = currentDocument(document, raw);
  if (current.hasHooksContainer && Object.keys(current.hooks).length === 0) {
    raw = changeJsonc(raw, ['hooks'], undefined);
  }
  return raw;
}

function appendGroup(
  document: QwenDocument,
  raw: string,
  event: ManagedQwenEvent,
  group: QwenGroup,
): string {
  const current = currentDocument(document, raw);
  if (current.hooks[event]) {
    return changeJsonc(raw, ['hooks', event, current.hooks[event].length], group, true);
  }
  return changeJsonc(raw, ['hooks', event], [group]);
}

export function renderQwenDocument(
  document: QwenDocument,
  agentId: string | undefined,
  additions: ReadonlyMap<ManagedQwenEvent, QwenGroup>,
): RenderedQwenDocument {
  let raw = removeManagedEntries(document, document.raw, agentId);
  for (const event of MANAGED_EVENTS) {
    const group = additions.get(event);
    if (group) raw = appendGroup(document, raw, event, group);
  }
  const current = currentDocument(document, raw);
  if (additions.size === 0 && document.ownedFile && Object.keys(current.root).length === 0) {
    return { document, changed: true, next: undefined };
  }
  return { document, changed: raw !== document.raw, next: raw };
}

export function qwenDocumentLabel(document: QwenDocument): string {
  return qwenSourceLabel(document.kind);
}

File: sdks/go/cmd/elydora/plugins/qwen_config.go
package plugins

import (
	"bytes"
	"fmt"
	"sort"
)

const (
	qwenSystemDefaultsKind = "system-defaults"
	qwenUserKind           = "user"
	qwenWorkspaceKind      = "workspace"
	qwenSystemKind         = "system"
)

type qwenDocument struct {
	kind               string
	filePath           string
	exists             bool
	raw                []byte
	snapshot           *managedFileSnapshot
	root               map[string]any
	hooks              qwenHookSettings
	hasHooksContainer  bool
	disableAllHooks    *bool
	folderTrustEnabled *bool
	ownedFile          bool
}

type qwenRenderedDocument struct {
	document *qwenDocument
	changed  bool
	next     []byte
	remove   bool
}

func qwenSourceLabel(kind string) string {
	switch kind {
	case qwenSystemDefaultsKind:
		return "Qwen Code system defaults"
	case qwenWorkspaceKind:
		return "Qwen Code workspace settings"
	case qwenSystemKind:
		return "Qwen Code system override settings"
	default:
		return "Qwen Code user settings"
	}
}

func qwenDocumentLabel(document *qwenDocument) string {
	if document == nil {
		return "Qwen Code settings"
	}
	return qwenSourceLabel(document.kind)
}

func readQwenFolderTrust(
	root map[string]any,
	label string,
) (*bool, error) {
	securityValue, exists := root["security"]
	if !exists {
		return nil, nil
	}
	security, ok := securityValue.(map[string]any)
	if !ok || security == nil {
		return nil, fmt.Errorf(`%s field "security" must be an object`, label)
	}
	folderTrustValue, exists := security["folderTrust"]
	if !exists {
		return nil, nil
	}
	folderTrust, ok := folderTrustValue.(map[string]any)
	if !ok || folderTrust == nil {
		return nil, fmt.Errorf(`%s field "security.folderTrust" must be an object`, label)
	}
	enabledValue, exists := folderTrust["enabled"]
	if !exists {
		return nil, nil
	}
	enabled, ok := enabledValue.(bool)
	if !ok {
		return nil, fmt.Errorf(
			`%s field "security.folderTrust.enabled" must be a boolean`,
			label,
		)
	}
	return &enabled, nil
}

func parseQwenDocument(
	kind string,
	exists bool,
	filePath string,
	raw []byte,
	snapshot *managedFileSnapshot,
) (*qwenDocument, error) {
	label := fmt.Sprintf("%s at %s", qwenSourceLabel(kind), filePath)
	root, err := decodeJSONCObject(raw, label, false)
	if err != nil {
		return nil, err
	}
	var disabled *bool
	if value, hasFlag := root["disableAllHooks"]; hasFlag {
		flag, ok := value.(bool)
		if !ok {
			return nil, fmt.Errorf(`%s field "disableAllHooks" must be a boolean`, label)
		}
		disabled = &flag
	}
	hooks := qwenHookSettings{}
	hooksValue, hasHooks := root["hooks"]
	if hasHooks {
		hooks, err = readQwenHookSettings(hooksValue, label+` field "hooks"`)
		if err != nil {
			return nil, err
		}
	}
	folderTrustEnabled, err := readQwenFolderTrust(root, label)
	if err != nil {
		return nil, err
	}
	return &qwenDocument{
		kind: kind, filePath: filePath, exists: exists,
		raw: append([]byte(nil), raw...), snapshot: snapshot, root: root,
		hooks: hooks, hasHooksContainer: hasHooks,
		disableAllHooks: disabled, folderTrustEnabled: folderTrustEnabled,
		ownedFile: kind == qwenUserKind && bytes.HasPrefix(raw, []byte(qwenOwnedFileMarker)),
	}, nil
}

func createQwenDocument(kind, filePath string) (*qwenDocument, error) {
	raw := []byte("{}\n")
	if kind == qwenUserKind {
		raw = []byte(qwenOwnedFileMarker + "\n{}\n")
	}
	return parseQwenDocument(kind, false, filePath, raw, nil)
}

func renderQwenDocument(
	document *qwenDocument,
	agentID, runtimeRoot string,
	additions map[string]map[string]any,
) (*qwenRenderedDocument, error) {
	if document == nil {
		return nil, fmt.Errorf("Qwen Code user settings are required")
	}
	label := fmt.Sprintf("%s at %s", qwenDocumentLabel(document), document.filePath)
	editor, err := newJSONCEditor(document.raw, label, false)
	if err != nil {
		return nil, err
	}
	removals := managedQwenRemovals(document.hooks, agentID, runtimeRoot)
	for _, event := range qwenManagedEvents {
		eventRemovals := make([]qwenManagedRemoval, 0)
		for _, removal := range removals {
			if removal.event == event {
				eventRemovals = append(eventRemovals, removal)
			}
		}
		sort.Slice(eventRemovals, func(left, right int) bool {
			return eventRemovals[left].groupIndex > eventRemovals[right].groupIndex
		})
		for _, removal := range eventRemovals {
			groupPath := []any{"hooks", event, removal.groupIndex}
			if removal.removeGroup {
				if err := editor.remove(groupPath); err != nil {
					return nil, err
				}
				continue
			}
			sort.Sort(sort.Reverse(sort.IntSlice(removal.handlerIndexes)))
			for _, handlerIndex := range removal.handlerIndexes {
				if err := editor.remove(appendQwenPath(groupPath, "hooks", handlerIndex)); err != nil {
					return nil, err
				}
			}
		}
		if len(eventRemovals) > 0 {
			current, parseErr := parseQwenDocument(
				document.kind,
				document.exists,
				document.filePath,
				editor.pack(),
				document.snapshot,
			)
			if parseErr != nil {
				return nil, parseErr
			}
			if groups, exists := current.hooks[event].([]any); exists && len(groups) == 0 {
				if err := editor.remove([]any{"hooks", event}); err != nil {
					return nil, err
				}
			}
		}
	}
	current, err := parseQwenDocument(
		document.kind,
		document.exists,
		document.filePath,
		editor.pack(),
		document.snapshot,
	)
	if err != nil {
		return nil, err
	}
	if current.hasHooksContainer && len(current.hooks) == 0 {
		if err := editor.remove([]any{"hooks"}); err != nil {
			return nil, err
		}
	}
	for _, event := range qwenManagedEvents {
		group, exists := additions[event]
		if !exists {
			continue
		}
		current, err = parseQwenDocument(
			document.kind,
			document.exists,
			document.filePath,
			editor.pack(),
			document.snapshot,
		)
		if err != nil {
			return nil, err
		}
		switch {
		case !current.hasHooksContainer:
			err = editor.addProperty(nil, "hooks", map[string]any{event: []any{group}})
		case current.hooks[event] != nil:
			err = editor.appendArray([]any{"hooks", event}, group)
		default:
			err = editor.addProperty([]any{"hooks"}, event, []any{group})
		}
		if err != nil {
			return nil, err
		}
	}
	next := editor.pack()
	nextDocument, err := parseQwenDocument(
		document.kind,
		document.exists,
		document.filePath,
		next,
		document.snapshot,
	)
	if err != nil {
		return nil, err
	}
	remove := len(additions) == 0 && document.ownedFile && len(nextDocument.root) == 0
	return &qwenRenderedDocument{
		document: document,
		changed:  remove || !bytes.Equal(next, document.raw),
		next:     next,
		remove:   remove,
	}, nil
}

func appendQwenPath(base []any, parts ...any) []any {
	path := make([]any, 0, len(base)+len(parts))
	path = append(path, base...)
	return append(path, parts...)
}
